internal/eventq/eventq.go

```diff
@@ -0,0 +1,163 @@
+/*
+   Copyright The containerd Authors.
+
+   Licensed under the Apache License, Version 2.0 (the "License");
+   you may not use this file except in compliance with the License.
+   You may obtain a copy of the License at
+
+       http://www.apache.org/licenses/LICENSE-2.0
+
+   Unless required by applicable law or agreed to in writing, software
+   distributed under the License is distributed on an "AS IS" BASIS,
+   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+   See the License for the specific language governing permissions and
+   limitations under the License.
+*/
+
+package eventq
+
+import (
+	"io"
+	"time"
+)
+
+type EventQueue[T any] struct {
+	events      chan<- T
+	subscriberC chan<- eventSubscription[T]
+	shutdownC   chan struct{}
+}
+
+type eventSubscription[T any] struct {
+	c      chan<- T
+	closeC chan struct{}
+}
+
+func (sub eventSubscription[T]) publish(event T) bool {
```

Comment:
`publish` will be blocked if `sub.c` is full. So one consumer may block the others from receiving updates. Is it acceptable?

Reply:
We could consider adding a log after a timeout if this condition is hit. Reliably and efficiently sending events to multiple subscribers can get very sophisticated. Much of that sophistication is probably unnecessary, but it would be good to have visibility into any case which is blocking.

My hypothesis is that if the subscriber is active, then the 100 event buffer is more than enough for it to keep up with the the flow of events. Previously there was a single 1000 event buffer that would fill up when there were no subscribers to pull them out of the channel, I don't believe there was a case where there was an active subscriber and the channel still filled up. It would be good to know that though.